### PR TITLE
port:audacity{,-devel} : upgrade to 2.2.2/2.2.2.333

### PR DESCRIPTION
**Pull request by RJVB**
port:audacity{,-devel} : upgrade to 2.2.2/2.2.2.333

### CODE REVIEW

File: audio/audacity/Portfile [hunk 1]
```diff
@@ -6,6 +6,14 @@ PortGroup           active_variants 1.1
 PortGroup           compiler_blacklist_versions 1.0
 PortGroup           github 1.0
 PortGroup           wxWidgets 1.0
+PortGroup           compiler_blacklist_versions 1.0
+PortGroup           active_variants 1.1
+
+platform darwin {
+    if {${os.major} < 13} {
+        PortGroup   cxx11 1.1
```
Comment (mojca) [inline audio/audacity/Portfile]: (Do you want to say that in order to compile this on linux you don't need C++11?)

You should not make this conditional. Just put the portgroup on top next to all others, no conditionals.

File: audio/audacity/Portfile [hunk 2]
```diff
@@ -57,7 +61,7 @@ platforms           darwin
 license             GPL-2+
 # openssl is a dependency of curl in cmake
 license_noconflict  openssl
-maintainers         {gmail.com:rjvbertin @RJVB} openmaintainer
+maintainers         gmail.com:rjvbertin openmaintainer
```
Comment (mojca) [inline audio/audacity/Portfile]: Please remove this change. The old value is correct.

File: audio/audacity/Portfile [hunk 3]
```diff
@@ -91,18 +92,20 @@ depends_build-append \
                     port:cmake \
                     port:pkgconfig \
                     port:python27
+depends_lib-append  port:freetype \
```
Comment (mojca) [inline audio/audacity/Portfile]: Why twice depends_lib-append? Remove the second one and just continue.

File: audio/audacity/Portfile [hunk 4]
```diff
@@ -119,19 +122,31 @@ platform darwin {
...
                     patch-libnyquist-symbol-visibility.diff \
                     patch-fix-casts.diff \
                     patch-add-MenusMac.diff \
                     patch-skip-gcc-version-test.diff \
-                    patch-fix-audiounits.diff \
-                    patch-temp-fixes.diff
+                    patch-fix-audiounits.diff
+    # strange C++ errors, why would upstream builds not be affected?
+    patchfiles-append \
+                    ${PPREFIX}patch-unordered_map-fixes.diff
```
Comment (mojca) [inline audio/audacity/Portfile]: I don't get this. Why the prefix in front of the patch name?

File: audio/audacity/files/patch-wxw-31x.diff [hunk 5]
```diff
@@ -37,3 +37,54 @@ index 508d8b8002dc18b17c8700f9eb7061c4178e0e6b..fca75e17da856475cb602d8d5040bcff
...
+-            auto stdstr = str.ToStdWstring(); // no allocations, a cheap fetch
+-            using Hasher = hash< decltype(stdstr) >;
+-            return Hasher{}( stdstr );
+-         }
+-      };
+-#ifdef __AUDACITY_OLD_STD__
+-   }
+-#endif
+-}
++// provided by WX 3.2
```
Comment (mojca) [inline audio/audacity/files/patch-wxw-31x.diff]: Please delete the code rather than commenting it out.

**Comment by RJVB** [comment]
A couple of specific answers below, but as a general remark I'd like to point out that I'm trying to upgrade a port that I'm already the (only) maintainer of. I would have done that without PR if I had commit rights, and in that case would have had few if any remarks about what I consider my internal shop keeping.
The points Mojca commented on are all there for a reason and shouldn't have any consequences outside of my personal workflow.

As to not adding my github handle: I decided a while back that my ports will contain only my maintainer identification(s) for me which are genuinely and globally useful. I'll reconsider that when MacPorts gets a central mapping between email and github handle; that's not something that should have to be repeated in each and every Portfile.
Before someone mentions it: yes, I know a github handle can be useful to @mention someone. But just how often do you have a reason to do that with someone who's handle you don't already know and/or who isn't already participating? I consider it rather unpolite to do this in random discussions which leads to the kind of semi-spam you see on Facebook and similar sites so I am not at all interested in making it too easy to @mention me.

>(Do you want to say that in order to compile this on linux you don't need C++11?)

Extending that reasoning would mean that on OS X 10.9 and newer we don't need C++11 to compile this? ;)

No, on 10.9+ we don't need the CXX11 PG, so there's no point to include it. Similarly, it's only needed on Mac or rather, last time I looked it was written with the assumption of being used only on Mac.

MacPorts doesn't officially support other platforms than Mac, so I see no justification to file a ticket against the CXX11 PG that it should only activate its payload on that platform.

>+    patchfiles-append \
>+                    ${PPREFIX}patch-unordered_map-fixes.diff
>
>I don't get this. Why the prefix in front of the patch name?

Follow the flow! There are 2 versions of the patch, one for each subport. Using a prefix is a trick I use in many parts where this is the case to limit the number of conditional patchfiles statements.

The better question here is why this patch is necessary for us and apparently not for upstream. I'm in the dark here, my usual C++ references don't even mention std::tr1::unordered_map .
The unpatched code is rejected both by clang++ and g++-mp-7. The 10.9 Xcode clang (3.5.x) and clang 5.0 give slightly different errors, and my GCC installation is probably still partly hardwired to use the libc++ headers which could explain why I'm getting the error there too. I've only had the time to build 2.2.2 against wxWidgets 3.2 on Linux which introduces relevant changes so I cannot yet say whether or not the patch is required there with wxWidgets 3.0.3. I can however compare libstdc++ and libc++ builds there which might help me answer why upstream builds haven't been running into issues. Maybe because they use wxWidgets 3.0.2?

Or, this is an issue on 10.9 and not on later OS X versions? I have to observe that the tr1/unordered_map header comes from the old GCC 4.2.1 C++ headerfiles on my system, i.e. it's not provided by clang.

**Comment by pmetzger** [comment]
Currently, the standard is indeed to have the maintainer line include the github handle of the contributor. This is rather important because it allows us to use a github-based workflow for comments on pull requests and the like. I think it is unlikely that the project would want to change that, as github has been rather productive for people, but you could always take that up with the developers mailing list.

**Comment by RJVB** [comment]
>This is rather important because it allows us to use a github-based workflow for comments on pull requests and the like.

Please explain.

Someone who creates a PR or leaves a comment on an existing PR or commit already provides his or her handle.

Far be it from me to suggest that github handles shouldn't be allowed or that no one should put them in their Portfiles. I would of course prefer if this kind of change wasn't made to my ports behind my back, but I don't really have any control over that, do I now.
But again, I do not appreciate being dragged into ("@mention"ed in) some random context, be it on github or Facebook. I have email for being alerted about things I might or ought to be interested in.

**Comment by mojca** [comment]
It's not about commits you make, but about commits others want to make to your ports.

If you volunteer to maintain ports you volunteer to review bug reports and pull requests. When someone opens a pull request you will get mentioned to give you a chance to review it and if you don't object to changes in a reasonable timeframe (officially 72 hours), it can be merged. Not being able to mention you means that nobody will manually start a mail client to write you an email to let you know that your port is in the list of 100 others to be changed.

Addition of github handled will be applied in a mass-change to all ports at some point in not too distant future, so your ports would be affected again in any case. I agree that the current situation with both email and handle being mentioned at all the places is suboptimal and that it would make more sense to have a separate database with mapping and only put one of the two in the port, but I guess pull requests for that are acceptable.

I see that new wxWidgets 3.0  is out, so it might make sense to test your port against that as well. I will probably prepare a pull request (there are a bunch of dependencies).

Weird constructs like `tr1::unordered_map` are from pre-C++11 times. If you are using C++11, the source code should not end up in the `tr1` scope.

**Comment by pmetzger** [comment]
> It's not about commits you make, but about commits others want to make to your ports.

Precisely.

**Comment by RJVB** [comment]
> I will probably prepare a pull request (there are a bunch of dependencies).

What does that mean? 
I'm in no hurry to spend more hours building a load of new dependencies only to find there's a reason why Audacity wants wx 3.0.2 and not a newer version...

>Weird constructs like `tr1::unordered_map` are from pre-C++11 times. If you are using C++11, the source code should not end up in the `tr1` scope.

Maybe that's why things break, because the code does end up there. Or maybe it actually uses std::unordered_map instead and the tr1 version only when the OLD_STD macro is set.

**Comment by RJVB** [comment]
>Precisely.

I only have openmaintainer ports, meaning I don't mind that others make commits to them (what choice do I have, not being able to commit changes to them myself...)

**Comment by pmetzger** [comment]
@RJVB Even for openmaintainer ports, it is generally considered important to be able to contact the maintainer for non-trivial changes, because the maintainer may know things others don't. It's just not required that we wait 72 hours even if changes are trivial (like just the version plus checksums.) Again it's therefore considered useful, given that we now work in Github, to be able to invoke people by their github handle.

**Comment by mojca** [comment]
Many patches apply with some fuzz which is something I'm not too eager to see. I'm not sure if that's simply due to the difference between the main and the devel version and the fact that same patches are used for both.

**Comment by RJVB** [comment]
Possibly, and probably in part also because it isn't always easy (nor worth the trouble) to get patches in a set to apply cleanly when they do so in isolation.

**Comment by mojca** [comment]
Btw: why are you blacklisting gcc? May that blacklisting be removed now that `cxx11 1.1` is included?

**Comment by RJVB** [comment]
> Btw: why are you blacklisting gcc?

Can't remember. It slipped in somewhere in https://trac.macports.org/ticket/53470, probably because I copy/pasted bits from a relevant fix as applied to another port.

> May that blacklisting be removed now that `cxx11 1.1` is included?

Probably. It seems unlikely anyhow that anyone would attempt to use the targeted old GCC versions to build Audacity nowadays.

**Comment by mojca** [comment]
While I was nitpicking about the main port, I have some real troubles applying patches to the `-devel` port. I tried building `audacity-devel +wx32`. Can you please double-check if that in fact works for you?

**Comment by mojca** [comment]
I would in fact be super grateful if you could put all the patches for the `-devel` version under that `devel` subfolder. Audacity is quite actively developed, so I try to set a goal of eventually getting rid of all the patches required for the devel version. If the patches are mixed, it makes it a lot more difficult to follow what was actually fixed upstream and it makes upgrading just one of the two versions much more tricky.

**Comment by RJVB** [comment]
It's my habit to do that, but only for patches that require it, i.e. when the same patch cannot be applied to or isn't required for both subports. Knowing myself, maintaining 2 slightly different versions of a same patch is not going to be a good idea.

I wouldn't count on many of my patches to be upstreamed anytime soon, except for those addressing obvious errors. I've made the Audacity Team aware of my changes very early on, and they have mostly been concentrating their active development elsewhere. The only patch I can think of that was incorporated is one for Linux, enabling the Mac theme there too.
A good deal of the patches is concerned with building Audacity "the MacPorts" way, which is obviously not going to be relevant anywhere else, and the team seem to have little interest in migrating away form their own hacked wxWidgets version anytime soon either.
All in all I have better things to do than lobbying to upstream my patches to what's only a side-project for me.

**Comment by RJVB** [comment]
> Can you please double-check if that in fact works for you?

Done. Is it possible that I had not copied all the correct patches from my own repository? Let me know which patches still pose problem for you.

**Comment by mojca** [comment]
After the last update it's just `patch-wxw-31x.diff`, while many other files result in lots of offsets and fuzz. Before that I stopped counting after four files.

**Comment by RJVB** [comment]
>After the last update it's just `patch-wxw-31x.diff`, while many other files result in lots of offsets and fuzz. Before that I stopped counting after four files.

path-wxw-31x.diff doesn't apply? I don't understand, I did a `port -v patch` from scratch using the updated portdir and that completed just fine (apart from possible offsets and fuzz, which I ignore).

**Comment by mojca** [comment]
Ok, I found the reason. There were multiple patches touching the same code in `src/MemoryX.h` and I made a minimal change in one of the patches while testing. I simplified that patch now to minimise interference, so I can test further.

**Comment by RJVB** [comment]
My bad, I didn't double-check the +wx32 variant while testing, did I mess up the order in which patches have to be applied (the 31x patch must be last)?

**Comment by RJVB** [comment]
BTW, a bunch of interesting commits landed in the upstream repo today, so I'll probably be pushing another upgrade to the -devel port later tonight.

**Comment by mojca** [comment]
No, it was my fault. I committed the update to master before I noticed your last comment about further updates. The port can still be updated later.

**Comment by mojca** [comment]
There's one problem with forcing the port to fetch from `git`. All the older machines are unable to fetch the sources due to issues with certificates.

**Comment by RJVB** [comment]
Why is that, can't those certificates be updated?

Next time I can comment-out or remote that line in the mainstream copy (commenting-out will make it less likely I forget it in a subsequent upgrade).

**Comment by mojca** [comment]
* https://trac.macports.org/ticket/53347
* https://trac.macports.org/ticket/51516

**Pull request closed without merging**